Add story image preview

luigieli opened the pull request:
Part of #1525 

This PR is a follow-up for https://github.com/lobsters/lobsters/issues/1525 based on the [suggestion in the issue](https://github.com/lobsters/lobsters/issues/1525#issue-2999750092):

>One improvement: the image for our stories is always the Lobsters logo (or missing). After discussion in https://github.com/lobsters/lobsters/issues/547 I'm comfortable introducing libvips. Let's extract the image for the story we're linking to and overlay the Lobsters logo in the lower-left corner. We'll have more variety in our feeds and hopefully a clear visual distinction to people who see both a direct link to a post and our link to our story of that post.

Using `libvips`, this PR adds a Lobsters logo overlay to the image extracted from the linked story by creating an async job to follow the same logic as avatar handling. Note that I am not updating the deployment scripts to [avatar caching](https://github.com/lobsters/lobsters/blob/14bc8599c91f6d3acdd1fd4ce7fca5b99fcc64bd/hatchbox/pre-build#L7-L8) caching in this PR.

The [opengraph.dev](https://opengraph.dev/) was used to make the comparison.
|Before|After|
|---|---
|<img width="1004" height="867" alt="image" src="https://github.com/user-attachments/assets/e4196056-2864-4b8e-8745-98e09ee3577c" />|<img width="1000" height="708" alt="image" src="https://github.com/user-attachments/assets/a8fc2d8d-a500-496e-aa5d-02e41094b48f" />|

pushcx commented:
Thanks for taking this on, it'll certainly make our bot feeds more interesting. And it's great to introduce libvips for long-desired features like #547 and #986.

I can see how the design of `Story.fetched_attributes` led you to put this in the Story model, but I don't want to extend the pattern of having a model making http requests. I'd rather this work by having the `StoryController#create` action enqueue a new `CreateStoryCard` job. Move this this code in there. Maybe that's even two jobs, one to fetch the existing attributes and image. Then, if it sees a value for the image, it can enqueue a second job to fetch their image and save it with our logo added.

A small design thing, please call this the 'card_image_url' rather than just the 'image_url'. I know OpenGraph has a more generic name, but I still hear everyone call it the 'card image' and I want the variable names to have more than just the type in there.

An even smaller design note, please make the Lobsters logo much smaller, like half the width and height it is now.

For CI, the brakeman failure is it warning about the potential to use user-supplied data in filenames. The risk there is that a malicious user could set `params[:short_id]` so the final path is `/home/deploy/lobsters/current/public/story_image/../../app/initializers/foo.rb` and then they get to upload code that'll be run the next time the server starts. So we always want to be very careful about where we get strings that can wind up in filenames. Run `brakeman -I` when you're sure it's secure and it'll walk you through the process of ignoring the warning and leaving a note to explain why the code is safe.

The test failure about libvips - I'm guessing that doesn't happen to your local dev? It's probably something in the [CI action](https://github.com/lobsters/lobsters/blob/main/.github/workflows/check.yml) that needs to include libvips in the container that runs the tests.

You're really working up and down every layer of our stack with this! No wonder it's gotten a little complicated. You're off to a great start on this feature.

pushcx commented on app/controllers/story_image_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      response.headers["Expires"] = 1.hour.from_now.httpdate
+      send_file cached_image, type: "image/png", disposition: "inline"
+    else
+      CreateStoryCardJob.perform_later(s)
```
```suggestion
```

I understand the goal here, that you want to fill the cache, but it looks like the job doesn't generate an image if the story doesn't have one (or it's too large, etc).

So then every time someone tries to load the image from us we'd enqueue a job that's going to hit the story URL again, seems like a bad behavior for a bot. We should trust that your PR queues the job every time a story URL is edited, and we can fix any missing paths if a prod bug becomes apparent.

pushcx commented on app/jobs/create_story_card_job.rb:
```diff
@@ -0,0 +1,98 @@
+
+    # If it's a PDF or too large, skip
+    return nil if response.body.length >= 5.megabytes
+    return nil if response["content-type"].to_s.match?(/pdf/)
```
Did you catch someone doing this in prod? What a funny bit of real-world complexity.

Maybe this check should flip to an allowlist of png/jpg/svg? Maybe if we're very lucky libvips exposes a list of formats/mime types it supports?

pushcx commented on app/jobs/create_story_card_job.rb:
```diff
@@ -0,0 +1,98 @@
+    # Try <meta property="og:image">
+    card_image_url = parsed.at_css("meta[property='og:image']")&.attributes&.[]("content")&.text
+
+    # Then try <meta name="image">
```
I appreciate that you thought to check a few places, the `&.` for handling the complexities of often-invalid pages, and the prudent `rescue` at the end. Good design!

pushcx commented on app/jobs/create_story_card_job.rb:
```diff
@@ -0,0 +1,98 @@
+    # don't want to fail the retrieval for this. Security risk is minimal.
+    s.ssl_verify = false
+    headers = {
+      "User-agent" => "#{Rails.application.domain} for job",
```
```suggestion
```

Oof, you revealed an ugly old bug here. We should be sending a better `User-agent` header. I'll fix this on `main` now.

pushcx commented on app/jobs/create_story_card_job.rb:
```diff
@@ -0,0 +1,98 @@
+    s.ssl_verify = false
+    headers = {
+      "User-agent" => "#{Rails.application.domain} for job",
+      "Referer" => Rails.application.domain
```
```suggestion
      "Referer" => Routes.story_short_id_url(@short_id)
```

Assumes this method is moved into the StoryImage class explained in another comment.

(God bless the HTTP typo that will never die.)

luigieli commented:
Hey @pushcx, thanks for the feedback. I've addressed what you pointed out:
- Added a `StoryImage` model to handle the logic of fetching and caching, the job and controller just delegate to the model.
- Enabled `SSL` verification and fixed the `Referer` header.
- Added a `content-type` allowlist based on [libvips docs](https://www.libvips.org/).
- Removed the redundant image fetching, it is only fetched when the story is created or the URL is edited.

pushcx commented:
Aha. And it's clearly cropping to the center of the image (perhaps even with clever 'center of interest' logic?) so we can't just put the logo in the upper-left and be done with it. Dang.

And of course other Fediverse clients might crop differently. Maybe we can add metadata to indicate the preferred cropping? I seem to recall that's in the UI for posting an image to Mastodon.

Whatever you find, please don't lose too much time to it. In the priority of things, this is a pretty minor visual glitch.
